# Prova di esame (laboratorio) 

La prova consiste in un insieme di quiz proposti attraverso un Jupyter Notebook (non sono ammessi altri strumenti software) da svolgere sui computer del laboratorio (non sono ammessi altri strumenti hardware). Durante la prova si può consultare il libro di testo, le slide usate a lezione, un manuale python (approvato dal docente) oppure le guide in linea suggerite da Jupyter Notebook. Non sono ammessi altri strumenti di consultazione. 

### Prima della prova

Specificare la informazioni richieste nelle celle sottostanti e salvare il notebook con il proprio nome.

### Durante la prova

Nelle celle markdown, cancellare il testo "YOUR ANSWER HERE" con la propria risposta. Nelle celle codice, rimuovere la linea di codice
```python
raise NotImplementedError()
```
e sostituirla con la propria implementazione.
Dopo ogni cella codice che deve contenere la risposta al quesito, esiste una cella codice con delle *asserzioni*, ossia delle istruzioni che verificano la correttezza della propria soluzione. (Queste asserzioni saranno usate per valutare la correttezza della risposta data e assegnare un punteggio.) E' possibile eseguire queste celle per verificare che la soluzione proposta sia corretta.


## Traccia

### Dimensioni della testa tra fratelli

#### Descrizione
Il dataset frets.csv ha 25 righe e 4 colonne, più una riga di intestazione e una colonna indice

I dati consistono in misurazioni della lunghezza e della larghezza dei capi di coppie di fratelli adulti in 25 famiglie campionate a caso. Tutte le misure sono espresse in millimetri.

#### Formato
Questo dataset contiene le seguenti colonne:

- l1: La lunghezza della testa del figlio maggiore.
- b1: La larghezza della testa del figlio maggiore.
- l2: La lunghezza della testa del secondo figlio.
- b2: La larghezza della testa del secondo figlio.

### Caricamento dati 

Utilizzando il modulo `csv` della libreria standard, caricare il dataset in una lista denominata `dataset_lst` che contiene tanti elementi quante sono i record del file csv, e ciascun elemento della lista corrisponde alla lista dei campi del record corrispondente. (Includere anche l'intestazione e la colonna indice.)

In [1]:
import csv 

with open('frets.csv') as frets:
    dataset_lst = list(csv.reader(frets))
    
dataset_lst

[['', 'l1', 'b1', 'l2', 'b2'],
 ['1', '191', '155', '179', '145'],
 ['2', '195', '149', '201', '152'],
 ['3', '181', '148', '185', '149'],
 ['4', '183', '153', '188', '149'],
 ['5', '176', '144', '171', '142'],
 ['6', '208', '157', '192', '152'],
 ['7', '189', '150', '190', '149'],
 ['8', '197', '159', '189', '152'],
 ['9', '188', '152', '197', '159'],
 ['10', '192', '150', '187', '151'],
 ['11', '179', '158', '186', '148'],
 ['12', '183', '147', '174', '147'],
 ['13', '174', '150', '185', '152'],
 ['14', '190', '159', '195', '157'],
 ['15', '188', '151', '187', '158'],
 ['16', '163', '137', '161', '130'],
 ['17', '195', '155', '183', '158'],
 ['18', '186', '153', '173', '148'],
 ['19', '181', '145', '182', '146'],
 ['20', '175', '140', '165', '137'],
 ['21', '192', '154', '185', '152'],
 ['22', '174', '143', '178', '147'],
 ['23', '176', '139', '176', '143'],
 ['24', '197', '167', '200', '158'],
 ['25', '190', '163', '187', '150']]

In [30]:
assert len(dataset_lst) == 26
assert len(dataset_lst[0]) == 5

### Array bidimensionale 

Definire un array numpy di nome `dataset_array` di due dimensioni, in cui ciascuna riga contiene un record del dataset (esclusa l'intestazione) e ciascuna colonna corrisponde a uno dei 4 campi del dataset (esclusa la colonna indice). Ciascun elemento dell'array conterrà il valore *intero* corrispondente nel dataset. 

In [3]:
import numpy as np

dataset_array = np.array(dataset_lst)[1 : , 1 :].astype('int64')

print(dataset_array)

[[191 155 179 145]
 [195 149 201 152]
 [181 148 185 149]
 [183 153 188 149]
 [176 144 171 142]
 [208 157 192 152]
 [189 150 190 149]
 [197 159 189 152]
 [188 152 197 159]
 [192 150 187 151]
 [179 158 186 148]
 [183 147 174 147]
 [174 150 185 152]
 [190 159 195 157]
 [188 151 187 158]
 [163 137 161 130]
 [195 155 183 158]
 [186 153 173 148]
 [181 145 182 146]
 [175 140 165 137]
 [192 154 185 152]
 [174 143 178 147]
 [176 139 176 143]
 [197 167 200 158]
 [190 163 187 150]]


In [4]:
assert dataset_array.shape == (25, 4)
assert dataset_array.dtype == "int64"

### Area di un'ellisse 

1. Dal modulo `math` importare la costante `pi` (corrisponde al valore π);
2. Definire la funzione `ellipsis_area(a , b)` che calcola l'area di un'ellisse con raggi `a` e `b` secondo la formula $$A=\pi \cdot a \cdot b$$ 

3. Se almeno uno dei parametri della funzione sono negativi, la funzione solleva l'eccezione `ValueError` con un messaggio di errore 

In [8]:
from math import pi

def ellipsis_area(a , b):
    
    if a < 0 or b < 0:
        raise ValueError('I due parametri non possono essere minori di 0')
        
    return pi * a * b


print(ellipsis_area(1 , 1))
print(ellipsis_area(2 , 1))
print(ellipsis_area(1 , 2))
print(ellipsis_area(2 , 2))

3.141592653589793
6.283185307179586
6.283185307179586
12.566370614359172


In [9]:
assert abs(ellipsis_area(0,0)) < 1e-6
assert abs(ellipsis_area(3,6)-18*3.142) < 1e-2
assert abs(ellipsis_area(5,3)-ellipsis_area(3,5)) < 1e-6

### Array di ovali 

Definire un array numpy bidimensionale di nome `ovals` con lo stesso numero di righe di `dataset_array` e due colonne. La prima colonna conterrà l'area dell'ellisse i cui *diametri* (attenzione a non confondere raggi e diametri!) corrispondono alle colonne `l1` e `b1`, la seconda colonna conterrà l'area dell'ellisse i cui diametri corrispondono alle colonne `l2` e `b2`. Utilizzare i dati presenti in `dataset_array` *senza usare cicli `for`*. Potete utilizzare direttamente la formula per il calcolo dell'area di un'ellisse senza utilizzare la funzione `ellipsis_area`.

In [10]:
ovals = np.array(
    [(ellipsis_area(elem[0] / 2 , elem[1] / 2) , ellipsis_area(elem[2] / 2 , elem[3] / 2)) for elem in dataset_array]
)

print(dataset_array)

[[191 155 179 145]
 [195 149 201 152]
 [181 148 185 149]
 [183 153 188 149]
 [176 144 171 142]
 [208 157 192 152]
 [189 150 190 149]
 [197 159 189 152]
 [188 152 197 159]
 [192 150 187 151]
 [179 158 186 148]
 [183 147 174 147]
 [174 150 185 152]
 [190 159 195 157]
 [188 151 187 158]
 [163 137 161 130]
 [195 155 183 158]
 [186 153 173 148]
 [181 145 182 146]
 [175 140 165 137]
 [192 154 185 152]
 [174 143 178 147]
 [176 139 176 143]
 [197 167 200 158]
 [190 163 187 150]]


In [11]:
assert ovals.shape == (25,2)
assert ovals.dtype == 'float64'

In [12]:
assert ovals.min() > 16000
assert ovals.max() < 26000

### Medie 

- *Utilizzando le universal functions (ufuncs)* calcolare la media sulle colonne di `ovals` e memorizzare il risultato in `ovals_means`. (Consultate la documentazione ufficiale di numpy per trovare la funzione corretta.)
- *utilizzando il broadcasting* sottrarre a ciascun elemento di `ovals` la media di colonna corrispondente e memorizzare il risultato in `ovals_residuals`.

In [13]:
ovals_means = np.mean(ovals , axis = 0)
ovals_residuals = ovals - ovals_means

print(ovals_means)
print(ovals_residuals)

[22082.85166469 21591.03533477]
[[ 1168.86096269 -1206.02600379]
 [  736.89197283  2404.44935335]
 [-1043.6056636     58.46503928]
 [  -92.48848772   409.53801832]
 [-2177.72061154 -2519.99713115]
 [ 3565.11075922  1330.02466582]
 [  183.18626763   643.58667101]
 [ 2518.17500741   971.88310331]
 [  360.68625256  3009.99133733]
 [  536.61544116   586.25260509]
 [  129.77919252    29.40530724]
 [ -954.85567113 -1502.12111139]
 [-1583.95960001   494.36101997]
 [ 1644.02685155  2453.92943765]
 [  213.03139784  1614.33880097]
 [-4544.12527786 -5152.65177486]
 [ 1655.807824    1117.96716171]
 [  268.00926928 -1481.70075914]
 [-1470.07686632  -721.43533697]
 [-2840.59666145 -3837.10985117]
 [ 1139.80123065   494.36101997]
 [-2540.57456303 -1040.30699131]
 [-2868.87099533 -1824.13435838]
 [ 3755.96251293  3227.54662859]
 [ 2240.92945573   439.38314853]]


In [14]:
assert ovals_means.shape == (2,)
assert abs(ovals_residuals.sum()) < 1e-6

### Zip 

Definire un oggetto zip di nome `ovals_pairs` che rappresenta la sequenza di coppie (o1, o2) essendo o1 e o2 elementi della stessa riga di `ovals_residuals`.

(Suggerimento: `ovals_pairs` è un generatore, pertanto è conveniente eseguire ogni volta la cella che lo definisce quando si vogliono testare le celle che lo utilizzano) 

In [15]:
ovals_pairs = zip(ovals_residuals[: , 0] , ovals_residuals[: , 1])

In [16]:
assert isinstance(ovals_pairs, zip)

### Visi diversi 

*Utilizzando le comprensioni di lista*, definire la lista `ovals_diffs` di valori Booleani, tali che un elemento della lista è True se i valori della corrispondente coppia di `ovals_pairs` hanno segno opposto (p.e. la coopia (2,-3) oppure (-3,2) corrisponde a True; la coppia (2,3) oppure (-2,-3) corrisponde a False)

In [17]:
ovals_diffs = [np.product(elem) < 0 for elem in ovals_pairs]

print(ovals_diffs)

[True, False, True, True, False, False, False, False, False, False, False, False, True, False, False, False, False, True, False, False, False, False, False, False, False]


In [18]:
assert len(ovals_diffs) == 25
assert ovals_diffs[0] and not ovals_diffs[1]

### Conversione in stringa 

A partire dall'array `dataset_array` definire una stringa di nome `chain` che concateni tutti i valori dell'array scandendolo per riga e separando i valori con ":". (La stringa inizierà con `191:155:179...`)

In [19]:
chain = ':'.join(str(elem) for elem in dataset_array.flat)

print(chain)

191:155:179:145:195:149:201:152:181:148:185:149:183:153:188:149:176:144:171:142:208:157:192:152:189:150:190:149:197:159:189:152:188:152:197:159:192:150:187:151:179:158:186:148:183:147:174:147:174:150:185:152:190:159:195:157:188:151:187:158:163:137:161:130:195:155:183:158:186:153:173:148:181:145:182:146:175:140:165:137:192:154:185:152:174:143:178:147:176:139:176:143:197:167:200:158:190:163:187:150


In [20]:
import re

re.fullmatch(r'(\d{3}:)*', chain)

### Espressioni regolari 

Utilizzando le espressioni regolari, estrarre da chain la sottostringa che inizia per `55`, prosegue per un numero indefinito di caratteri e termina con `88`. Memorizzare la sottostringa in `substr`.

In [ ]:
substr

print(s)

In [47]:
assert len(substr) == 222
assert substr in chain